codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoderForBuffer.java

```diff
@@ -0,0 +1,683 @@
+            return ctx.close();
+        }
+
+        @Override
+        public Future<Void> deregister() {
+            return ctx.deregister();
+        }
+
+        @Override
+        public Future<Void> write(Object msg) {
+            return ctx.write(msg);
+        }
+
+        @Override
+        public Future<Void> writeAndFlush(Object msg) {
+            return ctx.writeAndFlush(msg);
+        }
+
+        @Override
+        public Promise<Void> newPromise() {
+            return ctx.newPromise();
+        }
+
+        @Override
+        public Future<Void> newSucceededFuture() {
+            return ctx.newSucceededFuture();
+        }
+
+        @Override
+        public Future<Void> newFailedFuture(Throwable cause) {
+            return ctx.newFailedFuture(cause);
+        }
+    }
+
+    private static final class CompositeBufferCumulator implements Cumulator {
+        @Override
+        public Buffer cumulate(BufferAllocator alloc, Buffer cumulation, Buffer in) {
+            if (cumulation.readableBytes() == 0) {
+                cumulation.close();
+                return in;
+            }
+            CompositeBuffer composite;
+            try (in) {
+                if (CompositeBuffer.isComposite(cumulation)) {
+                    CompositeBuffer tmp = (CompositeBuffer) cumulation;
+                    // Since we are extending the composite buffer below, we have to make sure there is no space to
+                    // write in the existing cumulation.
+                    if (tmp.writerOffset() < tmp.capacity()) {
+                        composite = tmp.split();
+                        tmp.close();
+                    } else {
+                        composite = tmp;
+                    }
+                } else {
+                    composite = CompositeBuffer.compose(alloc, cumulation.send());
+                }
+                composite.extendWith((in.readOnly() ? in.copy() : in).send());
+                return composite;
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "CompositeBufferCumulator";
+        }
+    }
+
+    private static final class MergeCumulator implements Cumulator {
+        @Override
+        public Buffer cumulate(BufferAllocator alloc, Buffer cumulation, Buffer in) {
+            if (cumulation.readableBytes() == 0) {
+                // If cumulation is empty and input buffer is contiguous, use it directly
+                cumulation.close();
+                return in;
+            }
+            // We must close input Buffer in all cases as otherwise it may produce a leak if writeBytes(...) throw
+            // for whatever close (for example because of OutOfMemoryError)
+            try (in) {
+                final int required = in.readableBytes();
+                if (required > cumulation.writableBytes() || cumulation.readOnly()) {
+                    return expandCumulation(alloc, cumulation, in);
+                }
```

Comment:
How about calling `cumulation.ensureWritable(in.readableBytes(), cumulation.capacity(), true)` instead of this dance with `expandCumulation`?

Reply:
Good point.

Since this is also for the case when `cumulation` is read-only, `ensureWritable` will not always work but I can do that for cases when it can be expanded without a new buffer allocation.